apps/root/src/services/earnService.ts

```diff
@@ -389,4 +400,275 @@ export class EarnService extends EventsManager<EarnServiceData> {
+      token: {
+        ...sdkStrategyTokenToToken(
+          strategy.farm.asset,
+          `${strategy.farm.chainId}-${strategy.farm.asset.address}` as TokenListId,
+          {},
+          strategy.farm.chainId
+        ),
+        type: TokenType.ERC20_TOKEN,
+      },
+      amount,
+    });
+  }
+
+  setPendingTransaction(transaction: TransactionDetails) {
+    if (transaction.type !== TransactionTypes.earnDeposit && transaction.type !== TransactionTypes.earnIncrease) return;
+
+    const { typeData } = transaction;
+    let { positionId } = typeData;
+    const { strategyId } = typeData;
+
+    const userStrategies = [...this.userStrategies.filter((s) => s.id !== positionId)];
+
+    if (transaction.type === TransactionTypes.earnDeposit) {
+      const newEarnPositionTypeData = transaction.typeData;
+      positionId = `${transaction.chainId}-${strategyId}-${transaction.hash}` as SdkEarnPositionId;
+
+      const depositFee = this.allStrategies
+        .find((s) => s.id === strategyId)
+        ?.guardian?.fees.find((fee) => fee.type === FeeType.deposit);
+      const newUserStrategy = getNewEarnPositionFromTxTypeData({
+        newEarnPositionTypeData,
+        user: transaction.from as Address,
+        id: positionId,
+        transaction: transaction.hash,
+        depositFee: depositFee?.percentage,
+      });
+
+      userStrategies.push({ ...newUserStrategy, pendingTransaction: transaction.hash });
+    }
+
+    const existingStrategy = this.userStrategies.find((s) => s.id === positionId);
+    if (existingStrategy) {
+      existingStrategy.pendingTransaction = transaction.hash;
+      userStrategies.push(existingStrategy);
+    }
+
+    this.userStrategies = userStrategies;
+  }
+
+  handleTransactionRejection(transaction: TransactionDetails) {
+    if (transaction.type !== TransactionTypes.earnIncrease && transaction.type !== TransactionTypes.earnDeposit) return;
+
+    const { typeData } = transaction;
+    const { positionId, strategyId } = typeData;
+
+    let userStrategies;
+
+    switch (transaction.type) {
+      case TransactionTypes.earnDeposit:
+        userStrategies = [
+          ...this.userStrategies.filter((s) => s.id !== `${transaction.chainId}-${strategyId}-${transaction.hash}`),
+        ];
+        break;
+      case TransactionTypes.earnIncrease:
+        const userStrategy = this.userStrategies.find((s) => s.id === positionId);
+        userStrategies = [...this.userStrategies.filter((s) => s.id !== positionId)];
+
+        if (userStrategy) {
+          userStrategies.push({
+            ...userStrategy,
+            pendingTransaction: '',
+          });
+        }
+        break;
+      default:
+        userStrategies = [...this.userStrategies];
+        break;
+    }
+
+    this.userStrategies = userStrategies;
+  }
+
+  handleTransaction(transaction: TransactionDetails) {
+    if (transaction.type !== TransactionTypes.earnDeposit && transaction.type !== TransactionTypes.earnIncrease) {
```

Comment:
Oh yeah, this logic is what we need in DCA!

Reply:
Made it better still :D